### PR TITLE
noopener for target blank  Fixes #5700

### LINKED ISSUE #5700: Add rel=noopener to Link component for security
The [Link](https://github.com/mozilla/addons-frontend/blob/e036a8967ac4e0b4671ded3d0e3a27475b7a1808/src/amo/components/Link/index.js) component should add `rel="noopener noreferrer"` whenever it receives `target="_blank"` so that we do not [leak window objects to the target site](https://mathiasbynens.github.io/rel-noopener/). Since props are passed from `Button` to `Link`, this cover buttons as well.

Part of the patch should be cleaning up all the code which was manually setting `rel` properties.

**For QA**: please make sure that any link which opens in a new window still works as expected.

### PR DESCRIPTION
**Pull request by nabendu82**
`Fixes #5700`
Initial draft to add rel="noopener noreferrer" whenever it receives target="_blank"

### CODE REVIEW

File: src/amo/components/Link/index.js [hunk 1]
```diff
@@ -108,10 +112,24 @@ export class LinkBase extends React.Component {
     }
 
     return (
-      <Link {...customProps} to={linkTo}>
-        {children}
-        {needsExternalIcon ? <Icon name={iconName} /> : null}
-      </Link>
+      <Fragment>
```
Comment (willdurand) [inline src/amo/components/Link/index.js]: I don't think we need a `Fragment` here, you could keep the same code as before and spread a set of props like `customProps`:

```js
const linkProps = {
  ...customProps,
  target,
  rel: target === '_blank' ? 'noopener noreferrer' : undefined,
};

<Link {...linkProps} to={linkTo}>
  {children}
  {needsExternalIcon ? <Icon name={iconName} /> : null}
</Link>
```

Comment (nabendu82) [reply src/amo/components/Link/index.js]: Good solution.

Comment (willdurand) [reply src/amo/components/Link/index.js]: @nabendu82 do you need further information to update your PR?

File: src/amo/components/Link/index.js [hunk 2]
```diff
@@ -107,11 +113,17 @@ export class LinkBase extends React.Component {
...
     return (
-      <Link {...customProps} to={linkTo}>
-        {children}
-        {needsExternalIcon ? <Icon name={iconName} /> : null}
-      </Link>
+        <Link {...linkProps} to={linkTo}>
+          {children}
+          {needsExternalIcon ? <Icon name={iconName} /> : null}
+        </Link>
     );
```
Comment (willdurand) [inline src/amo/components/Link/index.js]: You should probably run `yarn prettier-dev`

File: src/amo/components/Link/index.js [hunk 3]
```diff
@@ -20,13 +20,17 @@ export class LinkBase extends React.Component {
     prependClientApp: PropTypes.bool,
     prependLang: PropTypes.bool,
     to: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
+    target: PropTypes.string,
+    rel: PropTypes.string
```
Comment (willdurand) [inline src/amo/components/Link/index.js]: I don't think you need this prop :)

File: tests/unit/amo/components/TestLink.js [hunk 4]
```diff
@@ -205,4 +205,10 @@ describe(__filename, () => {
 
     expect(root.find(Icon)).toHaveProp('name', `external-dark`);
   });
+
+  it('if the `target` is _blank', () => {
```
Comment (willdurand) [inline tests/unit/amo/components/TestLink.js]:
```
it('adds rel="noopener noreferrer" when target is "_blank"', () => {
```

File: src/amo/components/Link/index.js [hunk 5]
```diff
@@ -58,6 +58,8 @@ export class LinkBase extends React.Component {
       prependClientApp,
       prependLang,
       to,
+      target,
+      rel,
```
Comment (willdurand) [inline src/amo/components/Link/index.js]: This is probably not needed.

File: tests/unit/amo/components/TestLink.js [hunk 6]
```diff
@@ -205,4 +205,14 @@ describe(__filename, () => {
 
     expect(root.find(Icon)).toHaveProp('name', `external-dark`);
   });
+
+  it('adds rel="noopener noreferrer" when target is "_blank"', () => {
+    const root = render({
+      href: '/test',
+      target: '_blank',
+      rel: 'noopener noreferrer',
```
Comment (willdurand) [inline tests/unit/amo/components/TestLink.js]: I think you don't need this.

File: tests/unit/amo/components/TestLink.js [hunk 7]
```diff
@@ -205,4 +205,14 @@ describe(__filename, () => {
...
+    });
+
+    expect(root.find('a')).toHaveLength(1);
```
Comment (willdurand) [inline tests/unit/amo/components/TestLink.js]: That looks ok to me, but I'd rather test that the link component has `target` prop and its corresponding `rel` prop:

```js
const root = render({ href: '/', target: '_blank' });

expect(root.find('a')).toHaveProp('target', '_blank');
expect(root.find('a')).toHaveProp('rel', 'noopener noreferrer');
```

Comment (willdurand) [reply tests/unit/amo/components/TestLink.js]: You can also add another test case with `target: 'not-blank'`. You should expect the link to have a prop `target` but no `rel` prop.

Comment (nabendu82) [reply tests/unit/amo/components/TestLink.js]: Changed the test as above  -
```
  it('adds rel="noopener noreferrer" when target is "_blank"', () => {
    const root = render({ href: '/', target: '_blank' });
    
    expect(root.find('a')).toHaveProp('target', '_blank');
    expect(root.find('a')).toHaveProp('rel', 'noopener noreferrer');
  });
```

But giving error - 
screen shot 2018-08-22 at 1 13 51 pm

File: src/amo/components/Link/index.js (hunk 5 above)
Comment (nabendu82) [reply src/amo/components/Link/index.js]: If i remove target and rel from line 61, 62. Two of the below tests are failing.
screen shot 2018-08-22 at 1 25 14 pm

Comment (willdurand) [reply src/amo/components/Link/index.js]: Yep, `target` is required, but not `rel`

File: tests/unit/amo/components/TestLink.js (hunk 7 above)
Comment (shikha82) [reply tests/unit/amo/components/TestLink.js]: Should i remove **rel="noopener noreferrer"** from various Component files - 
```
PermissionsCard, ContributeCard, discoPurify, FooterBase, DiscoPaneBase 
```
and also the tests from test files -
```
test_utils, TestAddons, TestLink
```

Comment (nabendu82) [reply tests/unit/amo/components/TestLink.js]: Should i remove **rel="noopener noreferrer"** from various Component files - 
```
PermissionsCard, ContributeCard, discoPurify, FooterBase, DiscoPaneBase 
```
and also the tests from test files -
```
test_utils, TestAddons, TestLink
```

Comment (willdurand) [reply tests/unit/amo/components/TestLink.js]: @nabendu82 yes, you can remove rel="noopener noreferrer" from all components in `src/amo/` but I don't think you should change `discoPurify` and other files in `src/disco/`.

**Pull request merged**